Fix bug in ray.errors and update its default behavior

mitchellstern opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?

This fixes a bug in `ray.errors(include_cluster_errors=True)` arising from the use of `ray.JobID.nil()` rather than `None` in the call to `state.error_messages()`.

Additionally, this changes the signature from `ray.errors(include_cluster_errors=True)` to `ray.errors(all_jobs=False)`, switching the default behavior from returning all errors to only the current job's errors.

Finally, this unifies the return types of `ray.errors()` and `state.error_messages()` as suggested by @pcmoritz and updates the documentation accordingly.

**Old behavior (some console output omitted):**

Console 1:

```
>>> import ray
>>> ray.init()
>>> @ray.remote
... def h(): return 1 / 0
... 
>>> ray.get(h.remote())
>>> ray.errors(include_cluster_errors=False)
[{'timestamp': 1567092332.797547, 'type': 'task', 'message': '\x1b[36mray_worker:__main__.h()\x1b[39m (pid=3740, host=Mitchells-MBP.hsd1.md.comcast.net)\n  File "<stdin>", line 2, in h\nZeroDivisionError: division by zero\n'}]
>>> ray.errors()
[{'timestamp': 1567092332.797547, 'type': 'task', 'message': '\x1b[36mray_worker:__main__.h()\x1b[39m (pid=3740, host=Mitchells-MBP.hsd1.md.comcast.net)\n  File "<stdin>", line 2, in h\nZeroDivisionError: division by zero\n'}]
```

Console 2:

```
>>> import ray
>>> ray.init(redis_address='192.168.1.101:46533')
>>> ray.errors(include_cluster_errors=False)
[]
>>> ray.errors()
[]
```

**New behavior (some console output omitted):**

Console 1:

```
>>> import ray
>>> ray.init()
>>> @ray.remote
... def h(): return 1 / 0
...
>>> ray.get(h.remote())
>>> ray.errors()
[{'type': 'task', 'timestamp': 1567092017.747211, 'message': '\x1b[36mray_worker:__main__.h()\x1b[39m (pid=3652, host=Mitchells-MBP.hsd1.md.comcast.net)\n  File "<stdin>", line 2, in h\nZeroDivisionError: division by zero\n'}]
>>> ray.errors(all_jobs=True)
{'01000000': [{'type': 'task', 'timestamp': 1567092017.747211, 'message': '\x1b[36mray_worker:__main__.h()\x1b[39m (pid=3652, host=Mitchells-MBP.hsd1.md.comcast.net)\n  File "<stdin>", line 2, in h\nZeroDivisionError: division by zero\n'}]}
```

Console 2:

```
>>> import ray
>>> ray.init(redis_address='192.168.1.101:53710')
>>> ray.errors()
[]
>>> ray.errors(all_jobs=True)
{'01000000': [{'type': 'task', 'message': '\x1b[36mray_worker:__main__.h()\x1b[39m (pid=3652, host=Mitchells-MBP.hsd1.md.comcast.net)\n  File "<stdin>", line 2, in h\nZeroDivisionError: division by zero\n', 'timestamp': 1567092017.747211}]}
```

## Related issue number

N/A

## Linter

- [X] I've run `scripts/format.sh` to lint the changes in this PR.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/16638/
Test PASSed.

robertnishihara commented:
Good catch! Can you add a test so that we don't have a regression in this behavior?

One place to add it is somewhere in this test https://github.com/ray-project/ray/blob/38231907f348a8a5521fed0cf2613658a349601b/python/ray/tests/test_basic.py#L2524.

Also note that the linting is failing. You'll need to run `scripts/format.sh`.

robertnishihara commented:
Also, thanks for the detailed explanation in the PR description!

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/16655/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/16657/
Test PASSed.

mitchellstern commented:
I updated the name of the keyword argument from `include_cluster_errors` to `all_jobs`.

Additionally, since many of the tests were failing due to the `ray.errors()` now returning a dictionary mapping job IDs to lists of errors instead of a flat list of errors, I added a new utility function `flat_errors()` to `python/ray/tests/utils.py` and updated the tests to use this function instead.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/16742/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/16743/
Test PASSed.

richardliaw commented on python/ray/tests/test_failure.py:
```diff
@@ -165,7 +166,7 @@ def get_val(self):
             return 1
 
     # There should be no errors yet.
-    assert len(ray.errors()) == 0
```
this change is not needed right?

mitchellstern replied:
You're right, I reverted the changes to this file.

richardliaw commented on python/ray/state.py:
```diff
@@ -1185,19 +1185,22 @@ def available_resources():
 
 
-def errors(include_cluster_errors=True):
+def errors(all_jobs=True):
```
why default this to True?

richardliaw replied:
this can cause odd behaviors

mitchellstern replied:
That's the current default behavior, do you think it's something we should change?

mitchellstern replied:
Updated the default behavior to `all_jobs=False` as per our offline discussion.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/16808/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/16811/
Test PASSed.
